services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioning.java

```diff
@@ -22,29 +22,122 @@
 
 import javax.security.auth.x500.X500Principal;
 
+import org.eclipse.hono.deviceregistry.service.tenant.TenantInformationService;
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.service.credentials.CredentialsService;
 import org.eclipse.hono.service.management.OperationResult;
 import org.eclipse.hono.service.management.credentials.CredentialsManagementService;
 import org.eclipse.hono.service.management.credentials.X509CertificateCredential;
 import org.eclipse.hono.service.management.credentials.X509CertificateSecret;
 import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.eclipse.hono.util.CredentialsResult;
+import org.eclipse.hono.util.TenantObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import io.opentracing.Span;
+import io.opentracing.tag.Tags;
 import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
 
 /**
  * Helper to auto-provision devices.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/dev/concepts/device-provisioning/#automatic-device-provisioning">
+ *      Automatic Device Provisioning</a>
  */
 public final class AutoProvisioning {
+    private static final Logger LOG = LoggerFactory.getLogger(AutoProvisioning.class);
 
     private AutoProvisioning() {
     }
 
+    /**
+     * Auto-provision a device if auto-provisioning feature is enabled.
```

Comment:
FMPOV the fact that the device will be provisioned based on information form the client certificate in the clientContext should be described here explicitly.